src/main/java/org/elasticsearch/index/translog/fs/FsTranslog.java

```diff
@@ -189,168 +248,198 @@ public long ramBytesUsed() {
-        int deleted = 0;
-        try {
+    /** asserts that all files were closed, if not throws an {@link AssertionError} with details regarding the open files */
+    public static void assertAllClosed() {
+        if (ChannelReference.openedFiles == null) {
+            return;
+        }
+        RuntimeException exampleAllocator = null;
+        ArrayList<String> files = new ArrayList<>();
+        for (Map.Entry<String, Map<FsChannelReader, RuntimeException>> file : ChannelReference.openedFiles.entrySet()) {
+            files.add(file.getKey());
+            for (RuntimeException allocator : file.getValue().values()) {
+                if (exampleAllocator == null) {
+                    exampleAllocator = new RuntimeException(file.getKey() + " is still open", allocator);
+                } else {
+                    exampleAllocator.addSuppressed(allocator);
+                }
+            }
+        }
+        if (exampleAllocator != null) {
+            throw new AssertionError("some translog files are still open [" + Strings.collectionToCommaDelimitedString(files) + "]", exampleAllocator);
+        }
+    }
+
+    /** force close an open reference captured in assertion code * */
+    public static void assertForceCloseAllReferences() {
+        if (ChannelReference.openedFiles == null) {
+            return;
+        }
+        for (Map.Entry<String, Map<FsChannelReader, RuntimeException>> file : ChannelReference.openedFiles.entrySet()) {
+            IOUtils.closeWhileHandlingException(file.getValue().keySet());
+        }
+    }
+
+
+    /** gets a list of unreferenced files (only works if assertions are enabled, returns an empty array otherwise) */
+    public String[] getUnreferenced() throws IOException {
+        if (ChannelReference.openedFiles == null) {
+            return Strings.EMPTY_ARRAY; // not supported
+        }
+        ArrayList<String> result = new ArrayList<>();
+        try (ReleasableLock lock = writeLock.acquire()) {
             for (Path location : locations) {
                 try (DirectoryStream<Path> stream = Files.newDirectoryStream(location, TRANSLOG_FILE_PREFIX + "[0-9]*")) {
                     for (Path file : stream) {
-                        if (isReferencedTranslogFile(file) == false) {
-                            try {
-                                logger.trace("delete unreferenced translog file: " + file);
-                                Files.delete(file);
-                                deleted++;
-                            } catch (Exception ex) {
-                                logger.debug("failed to delete " + file, ex);
-                            }
+                        final long id = parseIdFromFileName(file);
+                        if (id < 0) {
+                            logger.trace("failed to extract translog id from [{}]", file);
+                        } else if (ChannelReference.openedFiles.containsKey(file.toString()) == false) {
```

Comment:
I'm consistent with the capturing? Also I don't know mockFS yet, rather do it in another change..